src/librustc_data_structures/indexed_set.rs

```diff
@@ -276,19 +322,200 @@ impl<'a, T: Idx> Iterator for Iter<'a, T> {
+    }
+
+    fn universe_size(&mut self) -> usize {
+        match *self {
+            HybridIdxSetBuf::Sparse(_, size) => size,
+            HybridIdxSetBuf::Dense(_, size) => size,
+        }
+    }
+
+    pub fn clear(&mut self) {
+        let universe_size = self.universe_size();
+        *self = HybridIdxSetBuf::new_empty(universe_size);
+    }
+
+    /// Returns true iff set `self` contains `elem`.
+    pub fn contains(&self, elem: &T) -> bool {
+        match self {
+            HybridIdxSetBuf::Sparse(sparse, _) => sparse.contains(elem),
+            HybridIdxSetBuf::Dense(dense, _) => dense.contains(elem),
+        }
+    }
+
+    /// Adds `elem` to the set `self`.
+    pub fn add(&mut self, elem: &T) -> bool {
+        match self {
+            HybridIdxSetBuf::Sparse(sparse, _) if sparse.len() < SPARSE_MAX => {
+                // The set is sparse and has space for `elem`.
+                sparse.add(elem)
+            }
+            HybridIdxSetBuf::Sparse(sparse, _) if sparse.contains(elem) => {
+                // The set is sparse and does not have space for `elem`, but
+                // that doesn't matter because `elem` is already present.
+                false
+            }
+            HybridIdxSetBuf::Sparse(_, _) => {
+                // The set is sparse and full. Convert to a dense set.
+                //
+                // FIXME: This code is awful, but I can't work out how else to
+                //        appease the borrow checker.
+                let dummy = HybridIdxSetBuf::Sparse(SparseIdxSetBuf::new(), 0);
+                match mem::replace(self, dummy) {
+                    HybridIdxSetBuf::Sparse(sparse, universe_size) => {
+                        let mut dense = sparse.to_dense(universe_size);
+                        let changed = dense.add(elem);
+                        assert!(changed);
+                        mem::replace(self, HybridIdxSetBuf::Dense(dense, universe_size));
+                        changed
+                    }
+                    _ => panic!("impossible"),
+                }
+            }
+
+            HybridIdxSetBuf::Dense(dense, _) => dense.add(elem),
+        }
+    }
+
+    /// Removes `elem` from the set `self`.
+    pub fn remove(&mut self, elem: &T) -> bool {
+        // Note: we currently don't bother going from Dense back to Sparse.
+        match self {
+            HybridIdxSetBuf::Sparse(sparse, _) => sparse.remove(elem),
+            HybridIdxSetBuf::Dense(dense, _) => dense.remove(elem),
+        }
+    }
+
+    /// Converts to a dense set, consuming itself in the process.
+    pub fn to_dense(self) -> IdxSetBuf<T> {
+        match self {
+            HybridIdxSetBuf::Sparse(sparse, universe_size) => sparse.to_dense(universe_size),
+            HybridIdxSetBuf::Dense(dense, _) => dense,
+        }
+    }
+
+    /// Iteration order is unspecified.
+    pub fn iter(&self) -> HybridIter<T> {
```

Comment:
Here again you could use `impl Trait`, though you have to pick from two different iterator types so it is a bit more annoying. There would be two options: to use the `Either` type or to use a single chained type. I tend to prefer the latter.

```rust
pub fn iter(&self) -> impl Iterator<Item = T> {
    let (sparse_iter, dense_iter) = match self {
        HybridIdxSetBuf::Sparse(sparse, _) => (Some(sparse.iter()), None),
        HybridIdxSetBuf::Dense(dense, _) => (None, Some(dense.iter()),
    }
    sparse_iter
        .into_iter()
        .flatten()
        .chain(dense_iter.into_iter().flatten())
}
```

Reply:
I prefer this because it ought to be mildly more efficient, since you kind of pick which iterator you will use just once, in the beginning, vs on every element. But your current setup is closer to EIther, which is more obvious. It would look like:

```rust
use either::Either;
pub fn iter(&self) -> impl Iterator<Item = T> {
    match self {
        HybridIdxSetBuf::Sparse(sparse, _) => Either::Left(sparse.iter()),
        HybridIdxSetBuf::Dense(dense, _) => Either::Right(dense.iter()),
    }
}
```

that assumes that `rustc_data_structures` already takes the [either crate](https://crates.io/crates/either) as a dependency, though. I know that some of our crates do.